src/main/java/org/prlprg/rds/RDSInputStream.java

```diff
@@ -24,114 +26,55 @@ public void close() throws IOException {
    * @return the next byte of data, or -1 if the end of the stream is reached.
    * @throws IOException if an I/O error occurs.
    */
-  public int readRaw() throws IOException {
-    return in.read();
+  public int readRaw(String desc) throws IOException {
+    var input = in.read();
+    logger.log(input, desc);
```

Comment:
If the logging message depends on the context where `readRaw` (or similar ones) is called, I would suggest to have the logger called not inside that function but at the same level of the call.
If the logging message is always the same/can be deducted from the function parameters, then it can be in the function.

Reply:
It depends on the context where it's called. I was alternating back and forth on whether to have the log on the same level as or within the input/output stream functions, but I see the issue with the coupling so I'll pull it out of the method.